### PR TITLE
API wrapper helpers

### PR DESCRIPTION
**Pull request by sebaspv**
## Summary
This PR adds support for streaming dialogue systems through async-friendly API wrapper functions.
## Rationale
The PR's main rationale is adding sample re-usable functions using `Stream`-friendly functions for future example demos and reusability in custom async-friendly dialogue systems.
## Testing
These functions have been tested locally, but there will be future implementations of unit tests for <s>all of them</s>. Groq sentence streaming test has been added.

### CODE REVIEW

File: iftk/__init__.py [hunk 1]
```diff
@@ -6,3 +6,6 @@
 from .channel import AsyncChannel, Channel, ChannelClosedEvent, DequeChannel
 from .pubsub import BaseSubscriber, PubSub, PubSubChannel, Subscriber
 from .system import System
+from .helpers.deepgram_stream_queue import DeepgramStream
+from .helpers.eleven_stream import ElevenStream
+from .helpers.groq_stream import GroqProcessor
```
Comment (chrisklaiber) [inline iftk/__init__.py]: Let's not include these helpers in __init__. The reason is that a consumer of the library may want to use some but not all helpers, and so we don't want to require the respective package installations for, e.g., the deepgram SDK.

File: iftk/helpers/deepgram_stream_queue.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
```
Comment (chrisklaiber) [inline iftk/helpers/deepgram_stream_queue.py]: Nice, thanks for including all the necessary copyright notices!

File: requirements.txt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+pyaudio
+deepgram-sdk
+python-dotenv
+websockets
+deepgram-sdk
+groq
+elevenlabs
```
Comment (chrisklaiber) [inline requirements.txt]: A few minor nits:

1) Let's rename this file to `requirements-dev.txt` ... after thinking about it, I imagine this will be more clear about our intent.

2) Please alphabetically sort these. Then it's easier for future edits to insert at the right place.

3) Please add version numbers (e.g., `groq==0.9.0`). Then it's easier for us to track down future breakages that can occur due to dependency updaes.

File: tests/helpers/test_groq_stream.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import sys
+import os
+import unittest
+from groq import AsyncGroq
+from dotenv import dotenv_values
```
Comment (chrisklaiber) [inline tests/helpers/test_groq_stream.py]: nit: sort and separate imports (here and elsewhere) per PEP8 guidelines. 

I'd recommend setting up `black` for formatting and `isort` (https://pycqa.github.io/isort/) to handle import sorting which should handle it for you. Both packages can also be listed in `requirements-dev.txt`

File: tests/helpers/test_groq_stream.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../..")))
```
Comment (chrisklaiber) [inline tests/helpers/test_groq_stream.py:13]: Does this work with context.py?
```suggestion
from .context import iftk
```

File: iftk/helpers/deepgram_stream_queue.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        "speech_final" message is detected from Deepgram.
+    """
+    extra_headers = {"Authorization": f"Token {key}"}
+    audio_queue = asyncio.Queue()
+
+    def callback(in_data, frame_count, time_info, status):
+        """A basic callback for Deepgram's stream."""
+        audio_queue.put_nowait(in_data)
+        return (None, pyaudio.paContinue)
+
```
Comment (chrisklaiber) [inline iftk/helpers/deepgram_stream_queue.py]: Because asyncio.Queue is not thread safe, and deepgram calls back on some thread, we'll need to put this back on the asyncio loop.

```suggestion
    loop = asyncio.get_running_loop()
    
    def callback(in_data, frame_count, time_info, status):
        """A basic callback for Deepgram's stream."""
        loop.call_soon_threadsafe(audio_queue.put_nowait, in_data)
        return (None, pyaudio.paContinue)

```

File: iftk/helpers/deepgram_stream_queue.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    async with websockets.connect(
+        WSS_URL,
+        extra_headers=extra_headers,
+    ) as ws:
+
+        async def keep_alive(websocket: websockets.WebSocketClientProtocol):
+            while True:
+                keep_alive_msg = json.dumps({"type": "KeepAlive"})
+                websocket.send(keep_alive_msg)
+                print("KeepAlive message sent")
```
Comment (chrisklaiber) [inline iftk/helpers/deepgram_stream_queue.py]: remove this debug print, or move it to a logger debug statement, e.g., 
```suggestion
                # above, after the imports (per https://docs.python.org/3/library/logging.html) : logger = logging.getLogger(__name__)
                logger.debug("KeepAlive message sent")
```

File: iftk/helpers/deepgram_stream_queue.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                await asyncio.sleep(3)
+
+        asyncio.create_task(keep_alive(websocket=ws))
```
Comment (chrisklaiber) [inline iftk/helpers/deepgram_stream_queue.py]: Since this runs forever, we should cancel the task at some point.

File: iftk/helpers/deepgram_stream_queue.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        async def microphone():
+            audio = pyaudio.PyAudio()
+            stream = audio.open(
+                format=FORMAT,
+                channels=CHANNELS,
+                rate=DEEPGRAM_RATE,
+                input=True,
+                frames_per_buffer=DEEPGRAM_CHUNK,
+                stream_callback=callback,
+            )
```
Comment (chrisklaiber) [inline iftk/helpers/deepgram_stream_queue.py]: Let's split out the microphone reading from the Deepgram stream. You can do it in a separate PR or in this one.

The idea is that the same microphone reading code could be reused for multiple ASR stages.

File: iftk/helpers/deepgram_stream_queue.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+            stream.stop_stream()
+            stream.close()
+
+        async def sender(ws):
+            try:
+                while True:
+                    data = await audio_queue.get()
+                    await ws.send(data)
+            except Exception as e:
+                raise e
```
Comment (chrisklaiber) [inline iftk/helpers/deepgram_stream_queue.py]: nit: I'm not sure the try/except is doing anything, it looks OK to remove.

File: iftk/helpers/deepgram_stream_queue.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+                if len(msg["channel"]["alternatives"][0]["transcript"]) > 0:
+                    if transcript:
+                        transcript += " "
+                    transcript += msg["channel"]["alternatives"][0]["transcript"]
+                    if msg["speech_final"]:
+                        yield transcript
+                        transcript = ""
+
+        microphone_task = asyncio.ensure_future(microphone())
+        sender_task = asyncio.ensure_future(sender(ws))
```
Comment (chrisklaiber) [inline iftk/helpers/deepgram_stream_queue.py]: Looks like create_task is sufficient. ensure_future only seems needed if the thing may already be a task, coroutine, future, etc.

```suggestion
        microphone_task = asyncio.create_task(microphone())
        sender_task = asyncio.create_task(sender(ws))
```

File: iftk/helpers/deepgram_stream_queue.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+            async for msg in ws:
+                msg = json.loads(msg)
+
+                if len(msg["channel"]["alternatives"][0]["transcript"]) > 0:
+                    if transcript:
+                        transcript += " "
+                    transcript += msg["channel"]["alternatives"][0]["transcript"]
+                    if msg["speech_final"]:
+                        yield transcript
+                        transcript = ""
```
Comment (chrisklaiber) [inline iftk/helpers/deepgram_stream_queue.py]: I think this should be out-dented a few in case "speech_final" is ever be set when the transcript len == 0.
```suggestion

                if msg["speech_final"] and transcript:
                    yield transcript
                    transcript = ""
```

File: iftk/helpers/deepgram_stream_queue.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+        microphone_task = asyncio.ensure_future(microphone())
+        sender_task = asyncio.ensure_future(sender(ws))
+
+        async for update in receiver(ws):
+            transcription_queue.put_nowait(update)
+
+        tasks = [microphone_task, sender_task]
+        for task in tasks:
+            if not task.done():
+                task.cancel()
```
Comment (chrisklaiber) [inline iftk/helpers/deepgram_stream_queue.py]: It's OK to cancel if it's already done/canceled.
```suggestion
            task.cancel()
```

File: iftk/helpers/eleven_stream.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+from typing import AsyncIterator
+from typing import Optional
+from elevenlabs.client import ElevenLabs
+
+
+async def ElevenStream(
+    sentences: AsyncIterator,
+    eleven_client: ElevenLabs,
+    messages: Optional[dict] = None,
```
Comment (chrisklaiber) [inline iftk/helpers/eleven_stream.py]: Let's remove this. The TTS does not require a concept of the LM's dialog. Leaving it out makes this more flexible (maybe no LM is used, or one with different roles).
```suggestion
```

File: iftk/helpers/eleven_stream.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import AsyncIterator
+from typing import Optional
+from elevenlabs.client import ElevenLabs
+
+
+async def ElevenStream(
+    sentences: AsyncIterator,
```
Comment (chrisklaiber) [inline iftk/helpers/eleven_stream.py]: typing nit:
```suggestion
    sentences: AsyncIterator[str],
```

File: iftk/helpers/eleven_stream.py [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+        sentences (AsyncIterator): A sentence AsyncIterator
+        eleven_client (ElevenLabs): The 11Labs third-party TTS client.
+        messages (dict, optional): The message history of the dialogue system.
+
+    Yields:
+        Iterator[AsyncIterator]: A bytes iterator for playing audio using 11Labs stream audio playing function.
+    """
+    full_response = ""
+    async for sentence in sentences:
+        if sentence != "":
```
Comment (chrisklaiber) [inline iftk/helpers/eleven_stream.py]: style nit: go ahead and use a simple boolean test to handle empty strings:
```suggestion
        if sentence:
```

File: iftk/helpers/eleven_stream.py [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+from typing import Optional
+from elevenlabs.client import ElevenLabs
+
+
+async def ElevenStream(
+    sentences: AsyncIterator,
+    eleven_client: ElevenLabs,
+    messages: Optional[dict] = None,
+    voice: Optional[str] = "Jessica",
+) -> AsyncIterator:
```
Comment (chrisklaiber) [inline iftk/helpers/eleven_stream.py]: Can we type hint the response with whatever type EleventClient returns?
```suggestion
) -> AsyncIterator[???]:
```

File: iftk/helpers/deepgram_stream_queue.py [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+CHANNELS = 1
+RATE = 44100
+CHUNK = 512
+DEEPGRAM_CHUNK = 8000
+DEEPGRAM_RATE = 16000
+SECONDS = 2
+WSS_URL = "wss://api.deepgram.com/v1/listen?endpointing=500&encoding=linear16&sample_rate=16000&channels=1&interim_results=false"
+
+
+async def DeepgramStream(key: str, transcription_queue: asyncio.Queue) -> None:
```
Comment (chrisklaiber) [inline iftk/helpers/deepgram_stream_queue.py]: naming nit: functions should be snake_case
```suggestion
async def deepgram_stream(key: str, transcription_queue: asyncio.Queue) -> None:
```

Also, this function should probably take an AsyncIterator of audio events and return an AsyncIterator of transcriptions. It's OK to do that in a future PR.

File: iftk/helpers/deepgram_stream_queue.py [hunk 19]
```diff
@@ -0,0 +1,99 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (chrisklaiber) [inline iftk/helpers/deepgram_stream_queue.py]: naming nit: let's call this file just `deepgram.py`

File: iftk/helpers/eleven_stream.py [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import AsyncIterator
+from typing import Optional
+from elevenlabs.client import ElevenLabs
+
+
+async def ElevenStream(
```
Comment (chrisklaiber) [inline iftk/helpers/eleven_stream.py]: 2 naming nits:

1) Let's call this file just `elevenlabs.py`

2) Functions should be snake_case
```suggestion
async def eleven_stream(
```

File: iftk/helpers/groq_stream.py [hunk 21]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import AsyncIterator
+import groq
+
+
+async def GroqProcessor(llm_stream: groq._client.AsyncStream) -> AsyncIterator:
```
Comment (chrisklaiber) [inline iftk/helpers/groq_stream.py]: As before, 2 naming nits:

1) Let's name this file just `groq.py`

2) Functions should be snake_case (also I renamed it `_stream` to match the naming used above, and `_sentence_` because it makes decisions about breaking up the sentences).
```suggestion
async def groq_sentence_stream(llm_stream: groq._client.AsyncStream) -> AsyncIterator:
```

File: tests/helpers/test_groq_stream.py [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+
+from iftk.helpers.groq_stream import GroqProcessor
+
+DOTENV = dotenv_values(".env")
+GROQ_API_KEY = DOTENV["GROQ_API_KEY"]
+messages = [{"role": "user", "content": "How are you doing?"}]
+model_id = "llama-3.1-70b-versatile"
+client = AsyncGroq()
```
Comment (chrisklaiber) [inline tests/helpers/test_groq_stream.py]: client creation should be within the test function. Either direct in the test, or, if needing to use it for every test, in the `setUp` method within `TestGroq`.

```
    def setUp(self):
        self.client = AsyncGroq()
```

File: tests/helpers/test_groq_stream.py [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+
+
+class TestGroq(unittest.IsolatedAsyncioTestCase):
+    async def test_sentences(self):
+        llm_stream = await client.chat.completions.create(
+            messages=messages, model=model_id, stream=True
+        )
+        async for sentence in GroqProcessor(llm_stream):
+            if sentence:
```
Comment (chrisklaiber) [inline tests/helpers/test_groq_stream.py]: Maybe we should filter out empty sentences within the groq_stream. What do you think?

Then we could assert them as non-empty here.

**Review by chrisklaiber — CHANGES_REQUESTED** [review]
Thanks for getting these integrated, and I like the in-line documentation you have added.

I left in a bunch of comments. If you find it easier in the future to one PR per helper, that's also fine.

File: tests/helpers/test_groq_stream.py (hunk 5 above)
Comment (sebaspv) [reply tests/helpers/test_groq_stream.py:13]: It might have something to do with relative imports and how `unittest` handles running multiple test files, but it doesn't seem to work even with other workarounds.

Comment (chrisklaiber) [reply tests/helpers/test_groq_stream.py:13]: OK, we can leave this as-is and figure it out later.

File: iftk/helpers/microphone.py [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import asyncio
+
+import pyaudio
+
+
+async def microphone_async(
```
Comment (chrisklaiber) [inline iftk/helpers/microphone.py]: 1) Let's rename this file to `pyaudio.py`

2) Let's just call this `microphone`. Since our library is async by default in most cases, we can generally leave off `_async` suffixes, I think.

File: iftk/helpers/microphone.py [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+        format=format,
+        channels=1,
+        rate=rate,
+        input=True,
+        frames_per_buffer=frames_per_buffer,
+        stream_callback=callback,
+    )
+
+    stream.start_stream()
+    while stream.is_active():
```
Comment (chrisklaiber) [inline iftk/helpers/microphone.py]: Curious, what makes it stop? Does the user cancel on the CLI somehow?

**Review by chrisklaiber — APPROVED** [review]
Nice work! I think this is good to go once the microphone.py renames are done. Approving ahead so you're unblocked to land once that is in.
